Python/optimizer_bytecodes.c

```diff
@@ -370,6 +370,27 @@ dummy_func(void) {
         res = sym_new_type(ctx, &PyUnicode_Type);
     }
 
+    op(_BINARY_OP_SUBSCR_TUPLE_INT, (tuple_st, sub_st -- res)) {
+        assert(sym_matches_type(tuple_st, &PyTuple_Type));
+        if (sym_is_const(ctx, sub_st)) {
+            assert(PyLong_CheckExact(sym_get_const(ctx, sub_st)));
+            long index = PyLong_AsLong(sym_get_const(ctx, sub_st));
+            assert(index >= 0);
+            int tuple_length = sym_tuple_length(tuple_st);
+            if (tuple_length == -1) {
+                // Unknown length
+                res = sym_new_not_null(ctx);
+            }
+            else {
+                assert(index < tuple_length);
+                res = sym_tuple_getitem(ctx, tuple_st, index);
+            }
+        }
+        else {
+            res = sym_new_not_null(ctx);
+        }
```

Comment:
@tomasr8, this can be improved to handle abstract tuples that may not be constant (and it cleans things up, since it includes a length check):

```suggestion
        long index = PyLong_AsLong(sym_get_const(ctx, 
sub_st));
        // ...PyLong_AsLong can fail, need to check for error here...
        res = sym_tuple_getitem(ctx, tuple_st, index)
```

Reply:
> this can be improved to handle abstract tuples that may not be constant

@brandtbucher, I think this already handles abstract tuples? Assuming that abstract means those that have `JIT_SYM_TUPLE_TAG` rather than `JIT_SYM_KNOWN_VALUE_TAG`? For tuples that just have `JIT_SYM_KNOWN_CLASS_TAG` we don't know the length so there's nothing we can do I believe.

But yeah, we can simplify the code since `_Py_uop_sym_tuple_getitem` can handle unknown length. The only difference is that it sets the type to `unknown` rather than `non null` as we do here. Is that a problem? Shouldn't `_Py_uop_sym_tuple_getitem` always return at least `JIT_SYM_NON_NULL_TAG`?

Reply:
Your current code is gated on `sym_is_const`, which abstract tuples fail. The rest of your understanding is correct, though.

> Is that a problem? Shouldn't `_Py_uop_sym_tuple_getitem` always return at least `JIT_SYM_NON_NULL_TAG`?

Sounds like you found your next PR. :)

Reply:
Ah, I completely misunderstood, sorry. I thought you were checking whether the tuple was const, not the index. Disregard!

Reply:
no worries :) I think we can still simplify the code a bit though!